### PR TITLE
[SPARK-27351][SQL] Wrong outputRows estimation after AggregateEstimation with only null value column

### PR DESCRIPTION
**Pull request by pengbo**
## What changes were proposed in this pull request?
The upper bound of group-by columns row number is to multiply distinct counts of group-by columns. However, column with only null value will cause the output row number to be 0 which is incorrect.
Ex:
col1 (distinct: 2, rowCount 2)
col2 (distinct: 0, rowCount 2)
=> group by col1, col2
Actual: output rows: 0
Expected: output rows: 2

## How was this patch tested?
According unit test has been added, plus manual test has been done in our tpcds benchmark environement.

### CODE REVIEW

**Comment by pengbo** [comment]
@wzhfy Can you please have a look?

Thanks

**Comment by srowen** [comment]
To paraphrase for someone who doesn't know this code well: normally there's no way to have >0 rows but 0 distinct rows... except when the column is all null? distinct would return 0 rows? then a groupBy on that column still returns 1 grouping for that column?

**Comment by cloud-fan** [comment]
> except when the column is all null? distinct would return 0 rows?

Seems like so. Can we check the null count to make sure it's an all-null column?

**Comment by pengbo** [comment]
>  when the column is all null? distinct would return 0 rows? then a groupBy on that column still returns 1 grouping for that column?

@srowen Thanks for your comment. 
Yes, that's actually what's happening. Currently if one group by col is all null, the group by output rows estimation will be always 0. Please recheck the example I provided, feel free to ask if you need more information.

**Comment by pengbo** [comment]
> > except when the column is all null? distinct would return 0 rows?
> 
> Seems like so. Can we check the null count to make sure it's an all-null column?

You mean define "all null value" is not only the distinct value equals to 0 but also with null count is greater than 0?

**Comment by pengbo** [comment]
@cloud-fan null value count check has been added as well, please recheck when available. thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala [hunk 1]
```diff
@@ -39,8 +39,16 @@ object AggregateEstimation {
       // Multiply distinct counts of group-by columns. This is an upper bound, which assumes
       // the data contains all combinations of distinct values of group-by columns.
       var outputRows: BigInt = agg.groupingExpressions.foldLeft(BigInt(1))(
-        (res, expr) => res *
-          childStats.attributeStats(expr.asInstanceOf[Attribute]).distinctCount.get)
+        (res, expr) => {
+          val columnStat = childStats.attributeStats(expr.asInstanceOf[Attribute])
+          val distinctValue: BigInt = if (columnStat.distinctCount.get == 0 &&
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: I'm slightly concerned about performance here. Can you save the value of distinctCount.get so it isn't accessed twice, or is it not actually recomputed twice as-is?

Comment (pengbo) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Done.
"distinctCount.get" is to get the value from the Option. The cost seems to be negligible, but it's better to save it by one variable.

**Review by attilapiros — COMMENTED** [review]
LGTM

**Review by srowen — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
+1, LGTM. Merged to master.

Thank you for your first contribution, @pengbo .
Also, thank you, @attilapiros and @srowen .

**Comment by gatorsmile** [comment]
FYI, the PR title is incomplete.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala [hunk 2]
```diff
@@ -39,8 +39,16 @@ object AggregateEstimation {
...
       var outputRows: BigInt = agg.groupingExpressions.foldLeft(BigInt(1))(
-        (res, expr) => res *
-          childStats.attributeStats(expr.asInstanceOf[Attribute]).distinctCount.get)
+        (res, expr) => {
+          val columnStat = childStats.attributeStats(expr.asInstanceOf[Attribute])
+          val distinctCount = columnStat.distinctCount.get
+          val distinctValue: BigInt = if (distinctCount == 0 && columnStat.nullCount.get > 0) {
+            1
+          } else {
+            distinctCount
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: If the nullCount is not empty, the value should be `distinctCount + 1`, right? 

@pengbo @dongjoon-hyun

**Comment by gatorsmile** [comment]
Thanks for your reviews!

A general comment about the code review. We should try our best to improve the test coverage for the unit tests. This PR basically exposes one of the scenarios we missed before. Thus, we should try to ask the contributors to improve the other similar cases too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala (hunk 2 above)
Comment (pengbo) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Good point.  
I will try and test it out. Another PR will be submitted if the problem exists.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Oh, right. @gatorsmile . Originally, this PR aims to fix the case of `column with only null value`, but that's another case which we should fix. Thanks!

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Oh, I inferred that the distinct count already correctly counted a null as a distinct value. If not, yeah, then distinctCount doesn't matter; it should add 1 iff nullCount is > 0. Agree, if this is broader, can we get an additional test of that case too?

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon .
Since you revert this, could you comment on here about the reason, too?

**Comment by dongjoon-hyun** [comment]
Since we revert this approach, we need to find another way to avoid `output rows: 0` situation.

**Comment by HyukjinKwon** [comment]
yup, sorry. It was reverted due to https://github.com/apache/spark/pull/24436#discussion_r287070146 reason.

**Comment by dongjoon-hyun** [comment]
@rxin , @gatorsmile , @cloud-fan , @HyukjinKwon .

Sorry, but I'm wondering if that is a correct reason for revert.
- First, this PR doesn't affect `SQL COUNT DISTINCT` result and `DataSet.stats` result at all.
- Second, this PR doesn't affect single column statistics.
- Although this is might be exposed to users, this is a meaningful fix for internal `AggregateEstimation` in Apache Spark CBO.

Do we really need to revert this to prevent mismatch with `Pandas` or other user facing SQL output?

**Comment by HyukjinKwon** [comment]
Sorry, I think I rushed to revert. It's 0 as distinct count everywhere for `ColumnStat`. I thought it's definitely a mistake cuz here we use it as 1.

https://github.com/apache/spark/blob/239082d9667a4fa4198bd9524d63c739df147e0e/sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala#L270
https://github.com/apache/spark/blob/b1857a4d7dfe17663f8adccd7825d890ae70d2a1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala#L50-L51

It's a bit late in my time. I will reread it closely tomorrow in KST.

**Comment by dongjoon-hyun** [comment]
Thanks for confirming. I'll create reverting PRs for more reviews.

**Comment by HyukjinKwon** [comment]
I am sorry @dongjoon-hyun. It's my big mistake I apolosise that I rushed. Let me open a PR to revert next time. Can you revert my revert right away or open a PR to revert mine?
It shouldn't say the row count is 0 cuz null can be grouped as well.

**Comment by HyukjinKwon** [comment]
I will revert my revert.

**Comment by dongjoon-hyun** [comment]
Never mind~ Please go ahead! Thank you.

**Comment by HyukjinKwon** [comment]
Sorry guys. it was my huge mistake. I will make sure we don't make such mistake next time.

**Pull request closed without merging**